Feature: uri-list support

sskorol opened the pull request:
See initial request #412 

Motivation: to avoid loading heavy attachments into report's context, it'd be nice to be able to attach just a list of URIs for further manual downloading.

Adding URI as a return type will affect core and require some effort on refactoring. 

Current approach assumes pure report face involvement. Implementation itself is based on `text/uri-list` mime type.

As input user could use the following syntax:
```java
	@Attachment(value = "{0}", type = "text/uri-list")
	public String saveLinks(final String attachName, final List<String> uris) {
		return uris.stream().collect(joining(System.lineSeparator()));
	}
```
As output there will be provided a table-based list of links.
<a href="http://www.imagecross.com/"><img src="http://hostingd.imagecross.com/image-hosting-08/4827uri-list.png"></a><br><a href="http://www.imagecross.com/">Image Hosting</a>

Click will either trigger a new tab opening (if file is already on server) or direct file downloading.

Changelog:
 * added new `text/uri-list` attachment type;
 * added corresponding dummy data;
 * increased amount of expected inputs in one of `AllureReportGenerator` tests;
 * added helper method for splitting an input string into URIs list according to specification rules.

sskorol commented:
@baev @just-boris any thoughts / comments on this? Would it work for #412 closing? Or it'd be better to leave it to end-user? E.g. via custom html attachments. However, it'll add additional complexity for additional styles and resources support / delivery, which would be inconvenient to the end-user.

just-boris commented:
I don't like this feature because it uses non-standard mime-type. All standard types are listed here: http://www.iana.org/assignments/media-types/media-types.xhtml

Also, I don't understand what issue you want to solve

lanwen commented:
You can use parameters to solve your issue - just add parameter event
https://github.com/allure-framework/allure-core/blob/master/allure-model/src/main/resources/allure.xsd#L95

is would show your links.

Your proposal about custom html attachment is a few lines of code to create wrapper - so this is preferable way for me.

sskorol commented:
@just-boris see discussion in #412. There were concerns about heavy files loading into reports' context.
But I agree with your point about non standard mime types.

sskorol commented:
@lanwen thanks, will check parameters

baev commented:
IMHO it is fine as far as `text/uri-list` is standard mime-type. Also, what about to support comments?

sskorol commented:
@baev according to `text/uri-list` mime type notes:

> Any lines beginning with the '#' character are comment lines and are ignored during processing. (Note that URIs may contain the '#' character, so it is only a comment character when it is the first character on a line.)

To handle comments I've added js helper for filtering lines, which start with `#`:
```javascript
export default function splitUriList(uri) {
    var os = require('os');
    return uri.split(os.EOL).filter(function (el) {
        return el.trim().length > 0 && !el.trim().startsWith('#');
    });
}
```

baev commented:
I mean maybe we could show the comments?

sskorol commented:
@baev ah, well, it won't be a problem to fix this, if required. The other question is: should we display comments on a separate line (as it is in original format) or inline them with related uri?

First case:
```text
# some comment
http://my.domain.com/log.txt
```
Second case:
```text
# some comment: http://my.domain.com/log.txt
```
At moment links are formatted as a table. So could be rendered like `new row` = `comment` then `next row` = `link`; or `new row` = `1st column` -> `comment`, `2nd column` -> `link`

baev commented:
I don't know, today is the first time I saw such files :smile:

sskorol commented:
@baev with comments printing it'll look like the following:

<a href="http://www.freeimagehosting.net/commercial-photography/california/palm-springs/"><img src="http://i.imgur.com/CVNwIr6.png" alt="Palm Springs commercial photography"></a>

baev commented:
@sskorol thanks for all the work!

just-boris commented on allure-report-face/src/helpers/splitUriList.js:
```diff
@@ -0,0 +1,6 @@
+    return uri.split(os.EOL).filter(function (el) {
+        return el.trim().length > 0 && !el.trim().startsWith('#');
+    });
+}
```
1) This directory is for Handlebars helpers, not for utility functions.
2) In browsers, you don't need to care about line endings.
You imported whole os library because of constant: https://github.com/substack/os-browserify/blob/master/browser.js#L45

just-boris replied:
@baev I will send a new PR with fixes

sskorol replied:
@just-boris @baev guys, changes from last screenshot haven't been pushed yet. Merged version doesn't support comments. @just-boris if you're going to create PR with fixes, please see the following commit, which adds comments support: https://github.com/allure-framework/allure-core/commit/24de705f310e52277c0789382071eb3064f851fa Would be appreciated, if you could add corresponding updates.

just-boris replied:
See PR #810
